internal/resolver/dns/dns_resolver.go

```diff
@@ -201,18 +207,49 @@ func (d *dnsResolver) Close() {
 
 func (d *dnsResolver) watcher() {
 	defer d.wg.Done()
+	backoffIndex := 1
+	nextBackoffRate := backoff.DefaultExponential.Backoff(backoffIndex)
 	for {
 		select {
 		case <-d.ctx.Done():
+			d.pollTimer.Stop()
 			return
 		case <-d.rn:
+			// Reset polling timer if active, as ResolveNow was called from ClientConn.
+			if !d.pollTimer.Stop() {
+				select {
+				case <-d.pollTimer.C:
+				default:
+				}
+			}
+			backoffIndex = 1
+			nextBackoffRate = backoff.DefaultExponential.Backoff(backoffIndex)
+		case <-d.pollTimer.C:
 		}
 
 		state, err := d.lookup()
 		if err != nil {
+			// Report error to the underlying grpc.ClientConn.
 			d.cc.ReportError(err)
 		} else {
-			d.cc.UpdateState(*state)
+			err = d.cc.UpdateState(*state)
+		}
+
+		if err == nil {
+			// Success; reset polling for next usage.
+			if !d.pollTimer.Stop() {
+				select {
+				case <-d.pollTimer.C:
```

Comment:
I'm not sure whether this is racy.  It _may_ be possible for `Stop` to return false, but not have _yet_ pushed to the channel.

However, considering the only way you can get here is with a stopped & drained timer channel (in `case <-d.rn` you stop&drain; in `case <-d.pollTimer.C`, it obviously fired & was drained by that case).  So, I don't think this is necessary at all.

Reply:
Deleted all these cases.